File: android/src/main/kotlin/com/example/geo_integrity_score/MonitoringPolicy.kt
package com.example.geo_integrity_score

import kotlin.math.max

data class MonitoringPolicy(
    val mode: String,
    val evaluationIntervalMillis: Long,
    val sensorWarmupMillis: Long,
    val autoStopOnIdle: Boolean,
    val speedThreshold: Double,
) {
    companion object {
        private const val DEFAULT_MODE = "balanced"
        private const val MIN_EVALUATION_INTERVAL_MS = 500L
        private const val DEFAULT_EVALUATION_INTERVAL_MS = 3_000L
        private const val DEFAULT_SENSOR_WARMUP_MS = 500L
        private const val DEFAULT_AUTO_STOP_ON_IDLE = true
        private const val DEFAULT_SPEED_THRESHOLD = 100.0

        fun fromMap(raw: Map<*, *>?): MonitoringPolicy {
            if (raw == null) {
                return balanced()
            }

            val mode = raw["mode"]?.toString() ?: DEFAULT_MODE
            val evaluationInterval = (raw["evaluationIntervalMillis"] as? Number)?.toLong()
                ?: DEFAULT_EVALUATION_INTERVAL_MS
            val warmup = (raw["sensorWarmupMillis"] as? Number)?.toLong()
                ?: DEFAULT_SENSOR_WARMUP_MS
            val autoStop = raw["autoStopOnIdle"] as? Boolean ?: DEFAULT_AUTO_STOP_ON_IDLE
            val speedThreshold = (raw["speedThreshold"] as? Number)?.toDouble()
                ?: DEFAULT_SPEED_THRESHOLD

            return MonitoringPolicy(
                mode = mode,
                evaluationIntervalMillis = max(MIN_EVALUATION_INTERVAL_MS, evaluationInterval),
                sensorWarmupMillis = max(0L, warmup),
                autoStopOnIdle = autoStop,
                speedThreshold = speedThreshold,
            )
        }

        fun aggressive(): MonitoringPolicy = MonitoringPolicy(
            mode = "aggressive",
            evaluationIntervalMillis = 1_000L,
            sensorWarmupMillis = 300L,
            autoStopOnIdle = false,
            speedThreshold = 120.0,
        )

        fun balanced(): MonitoringPolicy = MonitoringPolicy(
            mode = DEFAULT_MODE,
            evaluationIntervalMillis = DEFAULT_EVALUATION_INTERVAL_MS,
            sensorWarmupMillis = DEFAULT_SENSOR_WARMUP_MS,
            autoStopOnIdle = DEFAULT_AUTO_STOP_ON_IDLE,
            speedThreshold = DEFAULT_SPEED_THRESHOLD,
        )
    }
}

File: ios/Classes/MonitoringPolicy.swift
import Foundation

struct MonitoringPolicy {
  let mode: String
  let evaluationInterval: TimeInterval
  let sensorWarmup: TimeInterval
  let autoStopOnIdle: Bool
  let speedThreshold: Double

  static func from(_ arguments: Any?) -> MonitoringPolicy {
    guard let map = arguments as? [String: Any] else {
      return .balanced
    }

    let mode = (map["mode"] as? String) ?? "balanced"
    let evaluationMillis = (map["evaluationIntervalMillis"] as? NSNumber)?.doubleValue ?? 3_000
    let warmupMillis = (map["sensorWarmupMillis"] as? NSNumber)?.doubleValue ?? 500
    let autoStop = (map["autoStopOnIdle"] as? NSNumber)?.boolValue ?? (map["autoStopOnIdle"] as? Bool) ?? true
    let threshold = (map["speedThreshold"] as? NSNumber)?.doubleValue ?? 100

    return MonitoringPolicy(
      mode: mode,
      evaluationInterval: max(0.5, evaluationMillis / 1_000),
      sensorWarmup: max(0, warmupMillis / 1_000),
      autoStopOnIdle: autoStop,
      speedThreshold: threshold
    )
  }

  static var balanced: MonitoringPolicy {
    MonitoringPolicy(
      mode: "balanced",
      evaluationInterval: 3,
      sensorWarmup: 0.5,
      autoStopOnIdle: true,
      speedThreshold: 100
    )
  }
}
